drivers/watchdog/wdt_dw.c

```diff
@@ -0,0 +1,239 @@
+{
+	const struct dw_wdt_dev_cfg *const dev_config = dev->config;
+
+	/* Check component type */
+	if (dw_wdt_comp_type_get(dev_config->base) != WDT_COMP_TYPE_VALUE) {
+		LOG_ERR("Invalid component type %x", dw_wdt_comp_type_get(dev_config->base));
+		return -ENODEV;
+	}
+
+	dw_wdt_reset_pulse_length_set(dev_config->base, dev_config->reset_pulse_length);
+
+#if WDT_DW_INTERRUPT_SUPPORT
+	if (dev_config->irq_config)
+		dev_config->irq_config();
+#endif
+
+	return 0;
+}
+
+#if WDT_DW_INTERRUPT_SUPPORT
+static void dw_wdt_isr(const struct device *dev)
+{
+	const struct dw_wdt_dev_cfg *const dev_config = dev->config;
+	struct dw_wdt_dev_data *const dev_data = dev->data;
+
+	if (dw_wdt_interrupt_status_register_get(dev_config->base)) {
+		dw_wdt_clear_interrupt(dev_config->base);
+
+		if (dev_data->callback)
+			dev_data->callback(dev, 0);
+	}
+}
+#endif
+
+#define CHECK_RESET_PULSE_LENGTH_VALUE(rpl)                                                        \
+	(((rpl) == 2) || ((rpl) == 4) || ((rpl) == 8) || ((rpl) == 16) || ((rpl) == 32) ||         \
+	((rpl) == 64) || ((rpl) == 128) || ((rpl) == 256))
+
+#define CHECK_RESET_PULSE_LENGTH(inst)                                                             \
+	!CHECK_RESET_PULSE_LENGTH_VALUE(DT_INST_PROP_OR(inst, reset_pulse_length, 2)) ||
+
+#if DT_INST_FOREACH_STATUS_OKAY(CHECK_RESET_PULSE_LENGTH) 0
+#error Invalid reset pulse length value!
+#endif
+
+#define CHECK_CLOCK(inst)                                                                          \
+	!(DT_INST_NODE_HAS_PROP(inst, clock_frequency) || DT_INST_NODE_HAS_PROP(inst, clocks)) ||
+
+#if DT_INST_FOREACH_STATUS_OKAY(CHECK_CLOCK) 0
+#error Clock frequency not configured!
+#endif
+
+#define IRQ_CONFIG(inst)                                                                           \
+	static void dw_wdt##inst##_irq_config(void)                                                \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(inst), DT_INST_IRQ(inst, priority), dw_wdt_isr,           \
+			DEVICE_DT_INST_GET(inst), DT_INST_IRQ(inst, sense));                       \
+		irq_enable(DT_INST_IRQN(inst));                                                    \
+	}
+
+#define DW_WDT_INIT(inst)                                                                          \
+	IF_ENABLED(IS_INST_IRQ_EN(inst), (IRQ_CONFIG(inst)))                                       \
+                                                                                                   \
+	static const struct dw_wdt_dev_cfg wdt_dw##inst##_config = {                               \
+		.base = DT_INST_REG_ADDR(inst),                                                    \
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, clock_frequency),                          \
+			(.clk_freq = DT_INST_PROP(inst, clock_frequency)),                         \
+			(.clk_freq = DT_INST_PROP_BY_PHANDLE(inst, clocks, clock_frequency))       \
```

Comment:
Is this condition is really needed ? I saw that `clock_frequency` is not required property but I couldn't see `clocks` property

Reply:
This condition is necessary. The driver requires the `clock_frequency` or `clocks` property. `clocks` is not declared in `snps,designware-watchdog.yaml` because it is included from `base.yaml`. But there is a comment saying that this property is also supported.